Enable x86 support for Unix

am11 opened the pull request:
Also enabled cross compilation of x86 binaries on x64 host.

Ubuntu Dockerfiles:

<details>
  <summary><b><ins>Baseline: x64 build on x64 host</ins></b></summary>

```dockerfile
FROM ubuntu
RUN cat /etc/*-release

RUN apt-get update
RUN apt-get install -y \
        autoconf bash clang cmake gcc libtool curl \
        libunwind-dev llvm make openssl lldb git uuid-dev

RUN git clone https://github.com/dotnet/corert -b master --single-branch

WORKDIR /corert

RUN ./build.sh          # or ./build.sh x64

# usage:
# create a file name `Dockerfile` in an empty directory and run:
# docker build .
```
</details>

<details>
  <summary><b><ins>PR: x86 build on x86 host</ins></b></summary>

```dockerfile
FROM i386/ubuntu
RUN cat /etc/*-release

RUN apt-get update
RUN apt-get install -y \
        autoconf bash clang cmake gcc libtool curl \
        libunwind-dev llvm make openssl lldb git uuid-dev \
        g++-multilib

RUN git clone https://github.com/am11/corert -b linux-x86 --single-branch

WORKDIR /corert

RUN ./build.sh x86
```
</details>

<details>
  <summary><b><ins>PR: x86 build on x64 host</ins></b></summary>

```dockerfile
FROM ubuntu
RUN cat /etc/*-release

RUN apt-get update
RUN apt-get install -y \
        bash clang cmake gcc libtool curl \
        libunwind-dev llvm make openssl lldb git uuid-dev \
        g++-multilib
        
RUN git clone https://github.com/am11/corert -b linux-x86 --single-branch

WORKDIR /corert

RUN ./build.sh x86
```
</details>
<br />

Contributes to: #4589

am11 commented:
@jkotas, @janvorli this is WIP because around 100% of build, I get the assertion from libunwind on x86:

```sh
[ 97%] Building ASM object Runtime/Full/CMakeFiles/Runtime.dir/__/i386/PInvoke.S.o
[ 98%] Building ASM object Runtime/Full/CMakeFiles/Runtime.dir/__/i386/InteropThunksHelpers.S.o
[ 98%] Building ASM object Runtime/Full/CMakeFiles/Runtime.dir/__/i386/StubDispatch.S.o
[ 99%] Building ASM object Runtime/Full/CMakeFiles/Runtime.dir/__/i386/UniversalTransition.S.o
[100%] Building ASM object Runtime/Full/CMakeFiles/Runtime.dir/__/i386/WriteBarriers.S.o
In file included from /corert/src/Native/libunwind/src/libunwind.cpp:27:
In file included from /corert/src/Native/libunwind/src/UnwindCursor.hpp:30:
In file included from /corert/src/Native/libunwind/src/AddressSpace.hpp:35:
/corert/src/Native/libunwind/src/Registers.hpp:103:3: error: static_assert failed "x86 registers do not fit into unw_context_t"
  static_assert((check_fit<Registers_x86, unw_context_t>::does_fit),
  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /corert/src/Native/libunwind/src/libunwind.cpp:27:
/corert/src/Native/libunwind/src/UnwindCursor.hpp:616:3: error: static_assert failed "UnwindCursor<> does not fit in unw_cursor_t"
  static_assert((check_fit<UnwindCursor<A, R>, unw_cursor_t>::does_fit),
  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/corert/src/Native/libunwind/src/libunwind.cpp:66:24: note: in instantiation of member function 'libunwind::UnwindCursor<libunwind::LocalAddressSpace, libunwind::Registers_x86>::UnwindCursor'
      requested here
  new ((void *)cursor) UnwindCursor<LocalAddressSpace, REGISTER_KIND>(
                       ^
2 errors generated.
Runtime/Full/CMakeFiles/Runtime.dir/build.make:1526: recipe for target 'Runtime/Full/CMakeFiles/Runtime.dir/__/__/libunwind/src/libunwind.cpp.o' failed
make[2]: *** [Runtime/Full/CMakeFiles/Runtime.dir/__/__/libunwind/src/libunwind.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:107: recipe for target 'Runtime/Full/CMakeFiles/Runtime.dir/all' failed
make[1]: *** [Runtime/Full/CMakeFiles/Runtime.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
Failed to build corert native components.
```

I couldn't figure out the root cause. Could you please help?

jkotas commented on src/Native/Runtime/coreclr/gcinfotypes.h:
```diff
@@ -543,8 +542,8 @@ inline void GetInfoHdr(int index, InfoHdr * header)
 
-BYTE FASTCALL encodeHeaderFirst(const InfoHdr& header, InfoHdr* state, int* more, int *pCached);
-BYTE FASTCALL encodeHeaderNext(const InfoHdr& header, InfoHdr* state, BYTE &codeSet);
+Int8 FASTCALL encodeHeaderFirst(const InfoHdr& header, InfoHdr* state, int* more, int *pCached);
```
Could you please use standard C types for this: `int8_t`, etc.

am11 replied:
Fixed

janvorli commented:
You need to enlarge the unw_cursor_t buffer size. The size is defined here https://github.com/dotnet/corert/blob/a3b3fb0180970ff0bd410ae1c1f326b4e6341e49/src/Native/libunwind/include/__libunwind_config.h#L24

am11 commented:
@janvorli, thanks! Setting the values of cursor and context builds 100%. I found the correct values using sed:

```
for i in $(seq 1 100); do x="#  define _LIBUNWIND_CONTEXT_SIZE $i" ; sed -i "23s/.*/$x/" src/Native/libunwind/include/__libunwind_config.h ; ./build.sh x86 ; done
```

am11 commented:
@jkotas, this PR is ready. :)

The Jenkins-GH hooks seems to be broken as the legs are succeeded but here the CU status hasn't been updated.

jkotas commented:
@dotnet/dnceng The Windows CI machines picked up VS 2017, but they do not seem to have CMake that supports it:

```
12:26:52 [vcvarsall.bat] Environment initialized for: 'x86_x64'
12:26:53 CMake Error: Could not create named generator Visual Studio 15 2017 Win64
```

jkotas commented on src/Native/Runtime/PalRedhawkCommon.h:
```diff
@@ -134,6 +134,12 @@ struct PAL_LIMITED_CONTEXT
     UIntNative GetSp() const { return Rsp; }
     UIntNative GetFp() const { return Rbp; }
+
+#ifdef UNIX_X86_ABI
```
This ifdef is unnecessary.

am11 replied:
Removing this delta gives error on x86, adding SetIp and SetSp without ifdef statement will be unused on other architectures.

This is the error I received:

```sh
/corert/src/Native/Runtime/EHHelpers.cpp:401:15: error: no member named 'SetSp' in 'PAL_LIMITED_CONTEXT'
    pContext->SetSp(sp+sizeof(UIntNative)); // pop the stack
    ~~~~~~~~  ^
/corert/src/Native/Runtime/EHHelpers.cpp:445:21: error: no member named 'SetIp' in 'PAL_LIMITED_CONTEXT'
        palContext->SetIp((UIntNative)&RhpThrowHwEx);
        ~~~~~~~~~~  ^
```

jkotas replied:
It is fine to have the functions there, but have them unused in some build configs.

am11 replied:
Removed in the commit fixup.

jkotas commented on src/Native/Runtime/PalRedhawkCommon.h:
```diff
@@ -144,7 +150,7 @@ typedef void (__fastcall * ThreadExitCallback)();
 extern ThreadExitCallback g_threadExitCallback;
 
-typedef Int32 (*PHARDWARE_EXCEPTION_HANDLER)(UIntNative faultCode, UIntNative faultAddress, PAL_LIMITED_CONTEXT* palContext, UIntNative* arg0Reg, UIntNative* arg1Reg);
+typedef Int32 (__stdcall *PHARDWARE_EXCEPTION_HANDLER)(UIntNative faultCode, UIntNative faultAddress, PAL_LIMITED_CONTEXT* palContext, UIntNative* arg0Reg, UIntNative* arg1Reg);
```
It would be better to remove `__stdcall` from where this is used instead. The fewer places with the custom calling convention mods, the better.

am11 replied:
It is used in startup.cpp, but the actual issue is `PalRedhawh.h` has REDHAWK_PALIMPORT in its definition. Should I instead remove `REDHAWK_PALIMPORT` from the header?

jkotas replied:
I think removing it here https://github.com/dotnet/corert/blob/master/src/Native/Runtime/startup.cpp#L37 should be sufficient.

jkotas commented on src/Native/Runtime/coreclr/bitvector.h:
```diff
@@ -60,8 +59,9 @@
-    typedef UINT_PTR ChunkType;  // The size of integer type that the machine can operate on directly  
-//  typedef BYTE ChunkType;      // Use for testing
+    typedef uint32_t BOOL;
+    typedef unsigned int ChunkType;  // The size of integer type that the machine can operate on directly
```
This should be `uintptr_t` to match what it used to be

jkotas commented on src/Native/Runtime/coreclr/gcinfotypes.h:
```diff
@@ -543,8 +542,8 @@ inline void GetInfoHdr(int index, InfoHdr * header)
 
-BYTE FASTCALL encodeHeaderFirst(const InfoHdr& header, InfoHdr* state, int* more, int *pCached);
-BYTE FASTCALL encodeHeaderNext(const InfoHdr& header, InfoHdr* state, BYTE &codeSet);
+int8_t FASTCALL encodeHeaderFirst(const InfoHdr& header, InfoHdr* state, int* more, int *pCached);
```
This should be `uint8_t` to match `BYTE`.

markwilkie commented:
@crummel - looks like there's a CMake wrinkle...  #4637

crummel commented:
@mmitche should I update CMake on that same machine?

am11 commented:
Rebased on top of master wasm changes. Windows tests are passing locally with cmake 3.9.

jkotas commented on src/Native/Runtime/thread.h:
```diff
@@ -135,7 +135,7 @@ class Thread : private ThreadBuffer
     bool IsStateSet(ThreadStateFlags flags);
 
-    static UInt32_BOOL HijackCallback(HANDLE hThread, PAL_LIMITED_CONTEXT* pThreadContext, void* pCallbackContext);
+    static UInt32_BOOL _stdcall HijackCallback(HANDLE hThread, PAL_LIMITED_CONTEXT* pThreadContext, void* pCallbackContext);
```
I would rather remove `__stdcall` from both here and the header. Also, could you please fix the return type to UInt32 while you are on it to match the header?

jkotas commented:
@dotnet-bot test this please

jkotas commented:
LGTM modulo one nit.

@janvorli Could you please take a look as well?

janvorli commented on src/Native/Runtime/PalRedhawk.h:
```diff
@@ -367,7 +367,7 @@ typedef struct _CONTEXT {
     UIntNative GetSp() { return Esp; }
 } CONTEXT, *PCONTEXT;
-#include "poppack.h"
+#include "coreclr/poppack.h"
```
Please do not modify the includes of poppack.h / pshpackXXX.h, but rather add the path where they are located to the include paths in CMakeLists.txt.

janvorli commented on src/Native/Runtime/coreclr/poppack.h:
```diff
@@ -0,0 +1,40 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Why is this file placed to src/Native/Runtime/coreclr but the pshpack4.h to src/Native/Runtime?
It seems that a better place for all of these would be src/Native/inc/unix.

janvorli commented on src/Native/Runtime/thread.h:
```diff
@@ -135,7 +135,7 @@ class Thread : private ThreadBuffer
     bool IsStateSet(ThreadStateFlags flags);
 
-    static UInt32_BOOL HijackCallback(HANDLE hThread, PAL_LIMITED_CONTEXT* pThreadContext, void* pCallbackContext);
+    static UInt32 HijackCallback(HANDLE hThread, PAL_LIMITED_CONTEXT* pThreadContext, void* pCallbackContext);
```
Why have you changed the return type? The function returns TRUE / FALSE, so the return type was correct.

am11 replied:
This was to match typedef in the RedHawk.h header, where we have UInt32: https://github.com/dotnet/corert/pull/4637#discussion_r142313371.

janvorli replied:
Ah, I see. Ok, then I would rather fix it the other way - change the return type in the header to UInt32_BOOL. @jkotas since you have suggested the other way, are you ok with that?

jkotas replied:
Yes, fixing this other way is fine too.

am11 replied:
Thanks for the confirmation, I have fixed up all the feedback in 89a1373.

janvorli commented on src/Native/Runtime/unix/UnixContext.cpp:
```diff
@@ -361,10 +366,23 @@ void UnwindCursorToRegDisplay(unw_cursor_t *cursor, unw_context_t *unwContext, R
+    ASSIGN_REG(Ebp, Rbp)
+
+#define ASSIGN_TWO_ARGUMENT_REGS(arg0Reg, arg1Reg)    \
+    MCREG_Edi(nativeContext->uc_mcontext) = arg0Reg;  \
```
These should be ecx and edx unless we are going to use calling convention different from what we use in coreclr and on windows.

janvorli commented on buildscripts/build-managed.sh:
```diff
@@ -6,6 +6,8 @@ if [ "$BUILDVARS_DONE" != 1 ]; then
     . $scriptRoot/buildvars-setup.sh $*
 fi
 
+export __BuildArch=$__BuildArch
```
It seems that just `export __BuildArch` would be sufficient.

janvorli commented on src/Native/Runtime/i386/AllocFast.S:
```diff
@@ -0,0 +1 @@
+// TODO: Implement
```
A nit - could you please add the standard licence header to all of the new .S files?

am11 replied:
This is addressed, git diff is not showing here, full diff does: https://github.com/dotnet/corert/pull/4637/files#diff-240395baa0c54b3dd111b2a1f3645e4d

am11 commented:
Test OSX10.12 Release please
(intermittent segmentation fault: https://ci.dot.net/job/dotnet_corert/job/master/job/release_osx10.12_prtest/1005/consoleFull#11649539994b7730ef-150b-4be7-a00f-82ba6c50792b)

am11 commented:
@janvorli, I have addressed the feedback and also incorporated Alpine Linux / RID changes similar to dotnet/corefx#24423 here. Could you please take another look?

alpencolt commented:
@am11 we also interested in x86 support for Linux.
Are there any remained issues? Could share if you know.

am11 commented:
@alpencolt, we discussed the next steps a bit at https://github.com/dotnet/corert/issues/4589. Generally, I think there are at least two missing pieces:

* dotnet CLI for Linux and Windows x86 arch; which I haven't looked yet what is missing there
* ObjectWriter to support Linux and Windows x86 RID, I will try to build it first on Windows x86 without CoreCLR and LLVM dependencies. ObjectWriter project is currently located LLILC repo, ObjectWriter branch

alpencolt commented:
@am11 thank you.
Does the `ilc` compiler work now? I'm trying to compile simple program like:
```C#
private static int Main()
{
    return 99;
}
```
I do it on x86 chroot but it always failed with:
```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0xee7014d1 in ?? () from /home/u/src/arm_target_dir/libclrjitilc.so
[Current thread is 1 (Thread 0xf7375700 (LWP 15923))]
(gdb) bt
#0  0xee7014d1 in ?? () from /home/u/src/target_dir/libclrjitilc.so
#1  0x08ef5440 in ?? ()
#2  0xee8261d0 in ?? () at /home/alexander/src/corert/src/Native/jitinterface/jithost.cpp:41 from /home/u/src/target_dir/jitinterface.so
```
before JIT initialization. And it look it doesn't enter  to `JitHost::allocateMemory()`.
Have you face with such errors?

am11 commented:
@alpencolt, unfortunately I haven't gotten that far. I thought that we need dotnet-cli (specifically C# compiler csc) first on x86 to be able to compile the C# code. Right after 100% of build, the init-tools script fails with error that dotnet-sdk-linux-x86...tar not found. Could you please share the steps to compile C# on x86 system without having cli? :)

alpencolt commented:
@am11 I don't see messages about CLI, but I use cross build on x64.
Before building:
```
# To use RyuJIT compiler for building `Hello` program
touch tests/src/Simple/Hello/no_cpp
```
So at first you need to build CoreRT for x64, after that build CoreCLR, CoreFX and CoreRT for x86. I use for building the same commands for all projects:
```Shell
# Build rootfs only once
sudo ./cross/build-rootfs.sh x86 xenial
./build.sh clang3.9 x86 debug verbose cross skiptest
```

Install all necessary libraries by using script:
```Shell
WORKDIR=<YOUR_WORKING_DIR>
CORERT=$WORKDIR/corert
CORECLR=$WORKDIR/coreclr
COREFX=$WORKDIR/corefx
# ARM_DEST_DIR=$WORKDIR/arm_target_dir
ARM_DEST_DIR=$CORERT/arm_target_dir

# ARCH=arm
ARCH=x86
BUILD=Debug
BUILD_SUBPATH=Linux.${ARCH}.${BUILD}

rm -rf $ARM_DEST_DIR
mkdir -p $ARM_DEST_DIR

# I.   Copy CoreCLR(with CoreRun) part for our arch(arm)
cp ${CORECLR}/bin/Product/${BUILD_SUBPATH}/* ${ARM_DEST_DIR}
# Create copy libclrjit.so, this rename is hidden by the project make file in default building process.
# libclrjit.so is used by the CoreCLR that compiler runs on,
# and libclrjitilc.so is used for the actual compilation.
#cp ${ARM_DEST_DIR}/libclrjit.so ${ARM_DEST_DIR}/libclrjitilc.so for Legacy path
# cp ${ARM_DEST_DIR}/libprotojit.so ${ARM_DEST_DIR}/libclrjitilc.so
cp ${ARM_DEST_DIR}/libclrjit.so ${ARM_DEST_DIR}/libclrjitilc.so

# II.  Copy CoreRT part
# We use Hello sample to test and we need convert
# Hello.ilc.rsp absolute paths to relative for target
CORERT_ESCAPE=$(echo $CORERT | sed 's%/%\\/%g')
eval $(echo "cat ${CORERT}/tests/src/Simple/Hello/obj/${BUILD}/${ARCH}/native/Hello.ilc.rsp | sed 's/${CORERT_ESCAPE}\/tests\/src\/Simple\/Hello\/obj\/${BUILD}\/${ARCH}\//.\//' | sed 's/native\/Hello.o/Hello.o/' | sed 's/${CORERT_ESCAPE}\/tests\/..\/bin\/${BUILD_SUBPATH}\//.\//' > ${ARM_DEST_DIR}/Hello.ilc.rsp")

# Copy sample files and all needs librarires
CORERT_BUILD_SUBPATH=${BUILD_SUBPATH}
cp ${CORERT}/tests/src/Simple/Hello/obj/${BUILD}/${ARCH}/Hello.exe ${ARM_DEST_DIR}
echo "Skip cp ${CORERT}/libobjwriter.so ${ARM_DEST_DIR}!!!" #cp ${CORERT}/libobjwriter.so ${ARM_DEST_DIR}
cp ${CORERT}/bin/${CORERT_BUILD_SUBPATH}/tools/ilc.dll ${ARM_DEST_DIR}
cp ${CORERT}/bin/${CORERT_BUILD_SUBPATH}/tools/ILCompiler.* ${ARM_DEST_DIR}
cp ${CORERT}/bin/${CORERT_BUILD_SUBPATH}/tools/System.CommandLine.dll ${ARM_DEST_DIR}
cp ${CORERT}/bin/${CORERT_BUILD_SUBPATH}/tools/Microsoft.DiaSymReader.dll ${ARM_DEST_DIR}
cp ${CORERT}/bin/${CORERT_BUILD_SUBPATH}/tools/jitinterface.so ${ARM_DEST_DIR}
cp -r ${CORERT}/bin/${CORERT_BUILD_SUBPATH}/framework ${ARM_DEST_DIR}
cp -r ${CORERT}/bin/${CORERT_BUILD_SUBPATH}/sdk ${ARM_DEST_DIR}

# III. Copy CoreFX part
# Copy native architecture dependence libs(arm)
cp ${COREFX}/bin/${BUILD_SUBPATH}/native/* ${ARM_DEST_DIR}

NETCORE_PATH=netcoreapp-Linux-Debug-x64
# Copy arch independence
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Runtime.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Runtime.Extensions.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Collections.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Reflection.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Reflection.Metadata.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Collections.Immutable.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Console.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.IO.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Runtime.InteropServices.RuntimeInformation.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Runtime.InteropServices.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Threading.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Diagnostics.Debug.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Linq.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.IO.FileSystem.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.IO.MemoryMappedFiles.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.IO.UnmanagedMemoryStream.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.IO.FileSystem.Primitives.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Runtime.Handles.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Text.Encoding.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Text.Encoding.Extensions.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Reflection.Primitives.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Collections.Concurrent.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Security.Cryptography.Algorithms.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Security.Cryptography.Primitives.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Runtime.CompilerServices.Unsafe.dll ${ARM_DEST_DIR}
cp ${COREFX}/bin/runtime/${NETCORE_PATH}/System.Globalization.dll ${ARM_DEST_DIR}
```

Copy content of `${ARM_DEST_DIR}` to x86 chroot (it's located on `cross/rootfs/x86/`).
Enter to rootfs:
```
#!/bin/bash

CORERT_SRC=<PATH_TO_CORERT>
ARCH=x86
ROOT_FS=${CORERT_SRC}/cross/rootfs/${ARCH}

mount --bind /proc ${ROOT_FS}/proc
chroot ${ROOT_FS}
```

For translating program:
```
./corerun ilc.dll --verbose @Hello.ilc.rsp
```
And in the end link it on x64 host.
